dev-docs/RFCs/proposals/instance-culling-rfc.md

```diff
@@ -0,0 +1,178 @@
+# RFC: Instance Culling
+
+* **Author**: Ib Green
+* **Date**: Feb, 2021
+* **Status**: **Conceptual Draft**
+
+## Summary
+
+This RFC explores approaches to cull the number of instances to be rendered based
+on their distance or extents in view space, in order to increase rendering performance
+for large data sets. It also explores approaches to LOD (level of detail) to render
+simpler geometries for visually smaller objects.
+
+## Motivation
+
+The initial focus is on improving performance when using the `ScenegraphLayer`
+with high instance counts and high polygon count scene graphs.
+
+However the instance culling techniques can potentially be used to provide more 
+efficient filtering and rendering mechanisms for general layers.
+
+## Use Cases
+
+### SceneGraphLayer vertex explosion
+
+A use case for the scenegraph layer is to show objects on a map. The objects may be cars,
+drones, people etc.
+
+The `ScenegraphLayer` makes it possible to use high-quality, high-polygon count geometries.
+
+However the number of vertices (polygons) needed to be rendered quickly increases.
+
+For instance, a scene graph with 10K vertices, and 1000 instances, generates 10M vertices in a draw call.
+
+## Potential Approaches
+
+There are a number of techniques that can be applied to the problem:
+
+- **distance culling** - Determining the distance at which each instance is anchored
+- **bounding boxes** - Using bounding boxes to automatically determine distance culling limits
+- **vertex filtering** - A simpler, less performant approach of filtering out unnecessary work by filtering vertices
+- **instance filtering** - A more powerful, potentially more performant approach of filtering out unnecessary rendering by filtering the instance list 
+- **run-time tiling** - Separating a big table in sub-tables based on octree or runtime KD-tree decomposition, in the style of I3S or 3D tiles. 
+- **LOD** - using different geometries
+- **Decimation** - Automatic generation of smaller, reduced detail geometries by implementing decimation algorithms.
+
+## Distance Culling
+
+The idea is to select which instances to draw based on the distance of their center points from the viewer. Many layers already have distance points.
```

Comment:
- Instead of using the distance to center, it needs to use the distance to the bounding box. The size of a model could be significant depending on the camera position.
- Sorting data based on LOD so that we can still leverage some instanced-drawing techniques?

Reply:
Some benchmarking is needed to understand the perf impact of this. It would require projecting the anchor position and transforming the bounding box of every instance on render. Overhead increases linearly with data size.

Reply:
Great points, some thoughts...

> An issue with tiling is that it has to be regenerated when instance bounding box changes. This defeats the performance benefit of using e.g. sizeScale of the ScenegraphLayer.

To cover all cases yes, however:
- I believe the 95% use case (at least for `ScenegraphLayer`) does not involve globally and dynamically changing scale of models. I believe most users would happy give up that feature for the benefit of being able to render 10K high resolution models.
- I believe that doing culling and LOD selection on center points will give very close to desired results for 95% of scenegraphs. The point at which one switches LOD level is somewhat subjective in any case, and some variation should be acceptable.
- Perhaps we could simply document a limitation for instance culling: Dynamic `sizeScale` does not affect LOD selection? Or a prop could activate the more expensive behavior for those that really need it?

>  It would require projecting the anchor position and transforming the bounding box of every instance on render. 

Ideas:
- Could it be possible project the camera to the coordinate system of the positions, rather than the reverse? Then we only project as single point.
- Tile bounding boxes could perhaps be adjusted smartly with the global scale, rather than instance boxes. With the idea of only culling instances in tiles whose bounding boxes cross the LOD screen size limit, this would then be handled by just instance culling more (or fewer) tiles, but not all of them.

> Sorting data based on LOD so that we can still leverage some instanced-drawing techniques?

- Instanced drawing is still used, but we do have to "filter" the instance table when the camera changes. 
- LOD selection depends on camera position, I can't really see a way to pre-sort other than cutting into cubes or "tiles" as suggested..